File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionScaleTo.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionScaleTo extends SKAction {
    private float startX;
    private float startY;
    private float deltaX;
    private float deltaY;

    public SKActionScaleTo(float scaleX, float scaleY) {
        this.deltaX = scaleX;
        this.deltaY = scaleY;
    }

    @Override
    void computeStart() {
        startX = getParent().xScale;
        startY = getParent().yScale;
    }

    @Override
    void computeAction(long elapsed) {
        float newScaleX = getParent().xScale;
        float newScaleY = getParent().yScale;

        switch (timingMode) {
            case Linear: {
                newScaleX = SKEaseCalculations.linear(elapsed, startX, deltaX - startX, duration);
                newScaleY = SKEaseCalculations.linear(elapsed, startY, deltaY - startY, duration);
            }
            break;

            case EaseIn: {
                newScaleX = SKEaseCalculations.easeIn(elapsed, startX, deltaX - startX, duration);
                newScaleY = SKEaseCalculations.easeIn(elapsed, startY, deltaY - startY, duration);
            }
            break;

            case EaseOut: {
                newScaleX = SKEaseCalculations.easeOut(elapsed, startX, deltaX - startX, duration);
                newScaleY = SKEaseCalculations.easeOut(elapsed, startY, deltaY - startY, duration);
            }
            break;

            case EaseInEaseOut: {
                newScaleX = SKEaseCalculations.easeInOut(elapsed, startX, deltaX - startX, duration);
                newScaleY = SKEaseCalculations.easeInOut(elapsed, startY, deltaY - startY, duration);
            }
            break;

            default:
                break;
        }

        getParent().xScale = newScaleX;
        getParent().yScale = newScaleY;
    }

    @Override
    void computeFinish() {
        getParent().xScale = deltaX;
        getParent().yScale = deltaY;
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionScaleTo.kt
package br.com.insanitech.spritekit

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionScaleTo(private val deltaX: Float, private val deltaY: Float) : SKAction() {
    private var startX: Float = 0.toFloat()
    private var startY: Float = 0.toFloat()

    internal override fun computeStart() {
        startX = parent?.xScale ?: 0.0f
        startY = parent?.yScale ?: 0.0f
    }

    internal override fun computeAction(elapsed: Long) {
        var newScaleX = parent?.xScale ?: 0.0f
        var newScaleY = parent?.yScale ?: 0.0f

        when (timingMode) {
            SKActionTimingMode.Linear -> {
                newScaleX = SKEaseCalculations.linear(elapsed.toFloat(), startX, deltaX - startX, duration.toFloat())
                newScaleY = SKEaseCalculations.linear(elapsed.toFloat(), startY, deltaY - startY, duration.toFloat())
            }

            SKActionTimingMode.EaseIn -> {
                newScaleX = SKEaseCalculations.easeIn(elapsed.toFloat(), startX, deltaX - startX, duration.toFloat())
                newScaleY = SKEaseCalculations.easeIn(elapsed.toFloat(), startY, deltaY - startY, duration.toFloat())
            }

            SKActionTimingMode.EaseOut -> {
                newScaleX = SKEaseCalculations.easeOut(elapsed.toFloat(), startX, deltaX - startX, duration.toFloat())
                newScaleY = SKEaseCalculations.easeOut(elapsed.toFloat(), startY, deltaY - startY, duration.toFloat())
            }

            SKActionTimingMode.EaseInEaseOut -> {
                newScaleX = SKEaseCalculations.easeInOut(elapsed.toFloat(), startX, deltaX - startX, duration.toFloat())
                newScaleY = SKEaseCalculations.easeInOut(elapsed.toFloat(), startY, deltaY - startY, duration.toFloat())
            }

            else -> {
            }
        }

        parent?.xScale = newScaleX
        parent?.yScale = newScaleY
    }

    internal override fun computeFinish() {
        parent?.xScale = deltaX
        parent?.yScale = deltaY
    }

    internal override fun willHandleFinish(): Boolean {
        return false
    }
}
